### PR TITLE
Added downgrade method for forms and instances databases

### LINKED ISSUE #1452: Implement onDowngrade method for instances and forms databases
#### Problem description
Wee need to implement onDowngrade method for those databases since we will be extending them in the future and since we keep all data after downgrade.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #1452 

I decided to implement onDowngrade methods for forms and instances databases since these databases will be likely extended in the future so we really need that.

#### What has been done to verify that this works as intended?
I've tested upgrade/downgrade forms and instances databases using this scenario:
1. I installed this version 
2. Then I created new branch and added new columns for both tables.
3. I installed that version (upgrade went well)
4. I returned to this original branch and installed it again (downgrade went well)

**I was able to keep all data and have original tables in both cases.**  

#### Why is this the best possible solution? Were any other approaches considered?
Overall downgrading hasn't been supported in Android since 4.x? (I'm not sure about the version - it's not important). If you try to install an older app when you have a newer version of it you are asked to uninstall your current newer version first. Then all data is cleared (databases too) and that's why in most cases the problem with downgrading doesn't exist in other apps. In ODK Collect we store databases using external storage and we want to keep all data even if someone tries to downgrade (what might be dangerous - there were some related issues).

There is no a perfect way to restore old database and keep all data. 
When it comes to forms database I decided just to drop existed table and created new one (appropriate for new version). Then the database is auto filled using scanning forms functionality and it works well. Easy!
https://github.com/opendatakit/collect/pull/1453/files#diff-0c930d81466089bfec86df65dbef3a4eR80

When it comes to instances database it's not as easy as before because in instances table we have data like status for example (which can't be properly restored using scanning instances functionality). That's why I decided to recreate this table (like it was in the previous cases) but also read all data from original (old) table and try to insert to the new one. 
https://github.com/opendatakit/collect/pull/1453/files#diff-c6a9cd678a1a4f071894e0b4efe6cd92R161
This approach works well (I tested it) in described above scenario (if you only add/remove columns) but it's not possible to read/insert data if columns are renamed for example (so we should try to avoid such an operations). In the worst case if you rename any column and try to downgrade insert doesn't work and all data will be restored using scanning instances functionality.

To sum up: It should be a perfect solution when it comes to `forms.db` as we can restore all data scanning for forms. For `instnces.db` it should be also perfect solution assuming we don't implement changes like renaming columns (adding/removing columns is ok) in the worst case all data will be restored using scanning functionality too.

#### Are there any risks to merging this code? If so, what are they?
I think it's safe.

### CODE REVIEW

**Comment by grzesiek2010** [comment]
@mmarciniak90 I prepared for you two apk files. First one is an apk built using this branch and the second is an apk with newer database versions and new columns so you can easily test upgrade/downgrade.
[downgrade.zip](https://github.com/opendatakit/collect/files/1298944/downgrade.zip)

**Comment by mmarciniak90** [comment]
@grzesiek2010 thanks for sharing with me APKs! 
I finished quick review only on Android 6.0 and I did not notice unexpected behavior. 
Hope the same result will be on other Android versions. 

I noticed one disturbing behavior in other case. @grzesiek2010 could you look on this.
[apk.zip](https://github.com/opendatakit/collect/files/1299525/apk.zip)
I have installed this apk: **collectNewDatabase.apk**
Then I installed master apk :  **masterCollect-debug-v1.10.2-12-g4c0d0a2.apk**
When I open a ODK I get a error message:
screenshot_20170913-151155

**Comment by grzesiek2010** [comment]
@mmarciniak90 thanks for pointing this out. I've tested it carefully and it's an expected message.

For testing I prepared four branches:
**current_branch** - this branch
**branchX** - it's a branch which starts from **current_branch** and adds new columns (upgrade test)
**master_branch** - current master branch
**1.10.2_branch** - latest release branch

**Tested cases (installation order):**

| First version  | Second version | Third version | result |
| ------------- | ------------- | ------------- | ------------- |
| current_branch  | branchX | current_branch  | **OK**  |
| current_branch  | branchX  | 1.10.2_branch  | **OK**  |
| current_branch  | branchX  | master_branch  | **ERROR**  |

**Explanation:**
When we install the second version it's an upgrade and then it's a downgrade when we install the third version since only **branchX** has newer database version and new columns (for both forms and instances databases)

So if we merge this branch then the third (wrong case) will disappear.

The problem is because https://github.com/opendatakit/collect/pull/1363 has been already merged (**master_branch**) so it tries to call onDowngrade method which is not implemented yet (in ODKSQLiteOpenHelper we called only onUpgrade https://github.com/opendatakit/collect/pull/1363/files#diff-7f768ff010bb35b8c0a2a4a6fdbe6df5L128 and onDwongrade was skipped).

To sum up: It's an expected error since we need to add both #1334 and #1453 in one version. If we do that the third (wrong case) will disappear.

**Comment by mmarciniak90** [comment]
I noticed one more behavior that is worth to know.
**Use case**
1. User installs apk: _collect-debug-v1.10.2-5-g8f9ee1d.apk_
2. User finishes forms and sends one of them
3. User installs apk: _collectNewDatabase.apk_
4. User is able to see details of sent form
5. User installs again apk : _collect-debug-v1.10.2-5-g8f9ee1d.apk_
6. User opens 'View Sent Form' and user is not able to open details. 
screenshot_2017-09-14-14-52-10

**Workaround**
User need to open 'Fill Blank Form' view and then user will be able to see details of sent form.
ODK refreshes when user opens 'Fill Blank Form' view

In my opinion we don't need to fix this because there is a simple workaround and I believe this case is not common. I just wanted to inform about this behavior.

**Comment by grzesiek2010** [comment]
Yes that's because the forms database is recreated and I let to fill it in again using autoscan. 
Maybe it would be possible to run forms autoscan just after downgrade but I think the game is not worth the candle :smile: since it takes place only if someone tries to install an older version of the collect app and only if that newer one has upgraded forms database. So it will be rather not really common scenario and in that case we can use mentioned workaround)

**Comment by mmarciniak90** [comment]
Tested with success.
Verified on Android 4.1, 4.4, 5.1, 6.0, 7.0.
To test this issue I analyzed databases. 
1. I installed apk: _collect-debug-v1.10.2-5-g8f9ee1d.apk_ and downloaded one form. 
* I finished this form
* I sent one form
* I leaves one form as a incomplete

forms.db looked like this:


2. I installed apk: _collectNewDatabase.apk_
forms.db looked like this:

* new column 'upgradeDowngradeTest' is visible

3. I installed again apk : collect-debug-v1.10.2-5-g8f9ee1d.apk
forms.db looked like this:

* Column 'upgradeDowngradeTest' is no longer visible

For comparison, if apk with current master branch will be installed in point 3, the forms.db will look like this:

* Column 'upgradeDowngradeTest' is still visible.

File: collect_app/src/main/java/org/odk/collect/android/utilities/SQLiteQueryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.odk.collect.android.utilities;
+
+import android.database.sqlite.SQLiteDatabase;
+import android.database.sqlite.SQLiteException;
+
+public class SQLiteQueryBuilder {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/SQLiteQueryBuilder.java]: I think this name is confusing because it's the same as https://developer.android.com/reference/android/database/sqlite/SQLiteQueryBuilder.html

Honestly, I would tend to just do the queries inline so that it's super explicit what is going on. If you really want to introduce this abstraction, I think it should come with tests! It's relatively easy to test and it's also relatively easy to introduce really annoying bugs in something like this.

**Review by lognaturel — CHANGES_REQUESTED** [review]
Thanks for doing this, @grzesiek2010! It is indeed a good idea. I think we should also document what you're saying about renaming columns. I've made a note of it at https://docs.google.com/document/d/1dsmmPAa4t583kmhKB5IAGELHJFofnexOP9CGaaZwfnA/edit

File: collect_app/src/main/java/org/odk/collect/android/utilities/SQLiteQueryBuilder.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/utilities/SQLiteQueryBuilder.java]: I find it useful and clear so I added tests since I would like to keep it.

**Review by lognaturel — APPROVED** [review]
